docs/yaml/objects/compiler.yaml

```diff
@@ -659,3 +659,105 @@ methods:
       type: dep | list[dep]
       description: Additionally dependencies required.
       since: 1.1.0
+
+- name: symbol_visibility_header
+  returns: file
+  since: 1.3.0
+  posargs:
+    name:
+      type: str
+      description: Header file name.
+  kwargs:
+    namespace:
+      type: str
+      default: '<PROJECT_NAME>'
+      description: |
+        Prefix to use for `api`, `compilation` and `static_compilation` default
+        values. If those keyword arguments are defined, the namespace will not
+        be prepended.
+    api:
+      type: str
+      default: '<NAMESPACE>_API'
+      description: |
+        Name used for public APIs.
+    compilation:
+      type: str
+      default: '<NAMESPACE>_COMPILATION'
+      description: |
+        Name defined only when compiling the library.
+    static_compilation:
+      type: str
+      default: '<NAMESPACE>_STATIC_COMPILATION'
+      description: |
+        Name defined only when compiling or using static library.
+    static_only:
+      type: bool
+      default: false
+      description: |
+        If set to true, '<NAMESPACE>_STATIC_COMPILATION' is defined inside the
+        generated header.
+  description: |
+    Generate a header file that defines a macro to be used to mark all public APIs
+    of a library. Depending on the type of compiler and the platform, this will
```

Comment:
> Typically this would be `__declspec(dllexport)`, `__declspec(dllimport)` for MSVC

I think that's a common misconception to think this is a MSVC thing. This is a **Windows** thing, even with GCC. At least as far as I understood.

https://gcc.gnu.org/onlinedocs/gcc/Microsoft-Windows-Function-Attributes.html
> On systems that support the visibility attribute, this attribute also implies “default” visibility.

It implies default visibility, but is not synonym to it, that's what I understand from that doc at least.

Removed `the type of compiler and` in my sentence, because it depends more on the platform than compiler.

Reply:
I indeed thought so, but your sentence indeed cause me to doubt my recollections. I do almost no development on Windows. However, if the header does not depend on the compiler, someone could ask why this is a compiler object method, and we are back to the discussion about where this facility belongs...

The new sentence is clearer, but still gives the impression that different headers are generated on different platforms. I think it is important to communicate that the generated header is static. That's what my reformulation was trying to achieve.

Reply:
> think it is important to communicate that the generated header is static.

Added a sentence saying that explicitly.

> However, if the header does not depend on the compiler, someone could ask why this is a compiler object method, and we are back to the discussion about where this facility belongs...

It does not have to be a compiler method, but I personally think it's the most natural place. OTOH, it's not a hill I'm ready to die on, if @eli-schwartz really wants a module just for that, let's do it. Could be `meson.symbol_visibility_header()` too. But it's not clear he's going to accept it in that form, feels like he's going to reject it in any shape of form, but I could be wrong.

Reply:
I said I'd be happy with it as a module, and I mean it. :)

I do think that a module makes some sense, because there's other forms of snippets we may very well want too. Another possible use case is to generate test code snippets for some of the fancier autotools macros that projects sometimes find themselves using in order to test which of several different API conventions are used for the same meta-concept. Stuff like PAM_STRERROR_TWO_ARGS.

Why do you think that only visibility headers are going to end up being widely used/needed? :)

(This is also why I think the `meson` builtin object isn't any better of a location than the compiler object.)

Reply:
Just a drive-by note, potentially languages other than C may need visibility headers, which is why it makes sense to put it into a C compiler method.

Reply:
> Why do you think that only visibility headers are going to end up being widely used/needed? :)

Because I haven't seen requests for anything else yet, but I might very well just have missed them.

> potentially languages other than C may need visibility headers

Any specific language in mind? Currently this PR covers C/C++ and ObjC/ObjC++ which all use the same header file.